packages/flutter_tools/lib/src/android/android_device.dart

```diff
@@ -530,30 +533,39 @@ Map<String, String> parseAdbDeviceProperties(String str) {
   return properties;
 }
 
-// 015d172c98400a03       device usb:340787200X product:nakasi model:Nexus_7 device:grouper
-final RegExp _kDeviceRegex = new RegExp(r'^(\S+)\s+(\S+)(.*)');
-
 /// Return the list of connected ADB devices.
-///
-/// [mockAdbOutput] is public for testing.
-List<AndroidDevice> getAdbDevices({ String mockAdbOutput }) {
+List<AndroidDevice> getAdbDevices() {
+  final String adbPath = getAdbPath(androidSdk);
+  if (adbPath == null)
+    return <AndroidDevice>[];
+  final String text = runSync(<String>[adbPath, 'devices', '-l']);
   final List<AndroidDevice> devices = <AndroidDevice>[];
-  String text;
+  parseADBDeviceOutput(text, devices: devices);
+  return devices;
+}
 
-  if (mockAdbOutput == null) {
-    final String adbPath = getAdbPath(androidSdk);
-    printTrace('Listing devices using $adbPath');
-    if (adbPath == null)
-      return <AndroidDevice>[];
-    text = runSync(<String>[adbPath, 'devices', '-l']);
-  } else {
-    text = mockAdbOutput;
-  }
+/// Get diagnostics about issues with any connected devices.
+List<String> getAdbDeviceDiagnostics() {
+  final String adbPath = getAdbPath(androidSdk);
+  if (adbPath == null)
+    return <String>[];
+  final String text = runSync(<String>[adbPath, 'devices', '-l']);
+  final List<String> diagnostics = <String>[];
+  parseADBDeviceOutput(text, diagnostics: diagnostics);
+  return diagnostics;
+}
+
+// 015d172c98400a03       device usb:340787200X product:nakasi model:Nexus_7 device:grouper
+final RegExp _kDeviceRegex = new RegExp(r'^(\S+)\s+(\S+)(.*)');
 
+// Public for testing.
+void parseADBDeviceOutput(String text, {
```

Comment:
Please add docs on what the parameters are.

Reply:
added